python/src/iceberg/io/base.py

```diff
@@ -0,0 +1,123 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
```

Comment:
Might be worth including the explanation @rdblue gave on how these abstractions differ from other options, I can imagine it other's wondering about it.  One follow-up question I had is doesn't using the filesystem commitment protocol (atomic rename) require file listing?
```suggestion
Iceberg has its own FileIO abstraction instead of depending on an external library (e.g. fsspec) because Iceberg doesn't require directories and other features of a file system. It only cares that it can read a file as a stream with seek, or write a file and get the byte position from time to time. We therefore define our own abstractions that can wrap other file-like implementations.
```

Reply:
Atomicity is maintained at a higher level (specifically, the processes of swapping table metadata) so it doesn't depend on the filesystem interactions for atomic guarantees.

Reply:
Added a module level docstring that summarizes what Ryan said. 👍